Source/EasyNetQ/Internals/AsyncQueue.cs

```diff
@@ -0,0 +1,123 @@
+        ///     any release. You should only use it directly in your code with extreme caution and knowing that
+        ///     doing so can result in application failures when updating to a new EasyNetQ release.
+        /// </summary>
+        public AsyncQueue()
+        {
+        }
+
+        /// <summary>
+        ///     This is an internal API that supports the EasyNetQ infrastructure and not subject to
+        ///     the same compatibility as public APIs. It may be changed or removed without notice in
+        ///     any release. You should only use it directly in your code with extreme caution and knowing that
+        ///     doing so can result in application failures when updating to a new EasyNetQ release.
+        /// </summary>
+        public AsyncQueue(IEnumerable<T> collection)
+        {
+            foreach(var element in collection)
+                elements.Enqueue(element);
+        }
+
+        /// <summary>
+        /// Returns count of elements in queue
+        /// </summary>
+        public int Count
+        {
+            get
+            {
+                lock (mutex)
+                    return elements.Count;
+            }
+        }
+
+        /// <summary>
+        /// Dequeue element from queue
+        /// </summary>
+        /// <param name="cancellationToken">The cancellation token</param>
+        /// <returns>The dequeued element</returns>
+        public Task<T> DequeueAsync(CancellationToken cancellationToken = default)
+        {
+            lock (mutex)
+            {
+                if (elements.Count > 0)
+                    return Task.FromResult(elements.Dequeue());
+
+                FreeCancelledWaiters();
+
+                var waiter = new TaskCompletionSource<T>(TaskCreationOptions.RunContinuationsAsynchronously);
+                waiter.AttachCancellation(cancellationToken);
+                waiters.Enqueue(waiter);
+                return waiter.Task;
+            }
+        }
+
+        /// <summary>
+        /// Enqueue element to queue
+        /// </summary>
+        /// <param name="element">The element to enqueue</param>
+        public void Enqueue(T element)
+        {
+            lock (mutex)
+            {
+                var wasAttachedToWaiter = false;
+                while (waiters.Count > 0)
+                {
+                    var waiter = waiters.Dequeue();
+                    if (waiter.TrySetResult(element))
+                        wasAttachedToWaiter = true;
+                }
+
+                FreeCancelledWaiters();
+
+                if (wasAttachedToWaiter)
+                    return;
+
+                elements.Enqueue(element);
+            }
+        }
+
+        private void FreeCancelledWaiters()
+        {
+            while (waiters.Count > 0)
+            {
+                var waiter = waiters.Peek();
+                if (waiter.Task.IsCompleted)
```

Comment:
Method name is `FreeCANCELLEDWaiters`, but you use `IsCompleted` which means `RanToCompletion`, `Faulted` **OR** `Canceled`. Also I do not understand the purpose of `while` loop and `Peek`. In the case of pending waiters, this loop will run forever.

Reply:
>Method name is FreeCANCELLEDWaiters, but you use IsCompleted which means RanToCompletion, Faulted OR Canceled

Actually, there is only one way of completion of waiters which is still in waiters queue and it is their cancellation.

Reply:
>Also I do not understand the purpose of while loop and Peek. In the case of pending waiters, this loop will run forever.

You are right again. Thanks.

Reply:
>Also I do not understand the purpose of while loop and Peek

The only purpose of this loop is to clean up already cancelled waiters.

Reply:
Fixed.

Reply:
> Actually, there is only one way of completion of waiters which is still in waiters queue and it is their cancellation.

I agree but often there remains an unforeseen case when something goes wrong. I don’t know what’s best to do - whether to rename the method or just place your comment in code. I'm OK with current naming. It’s just that during the code review this section caused questions and it took some time to understand how everything works.

Reply:
A comment in any case doesn’t hurt.